Intra Rustdoc Links

killercup opened the pull request:
Add a notation how to create relative links in documentation comments (based on Rust item paths) and extend Rustdoc to automatically turn this into working links.

The important bits:

## Additions to the documentation syntax

1. `[Iterator](std::iter::Iterator)`
2. `[Iterator][iter]`, and somewhere else in the document: `[iter]: std::iter::Iterator`
3. `[Iterator]` or ```[`Iterator`]``` with automatically generated link reference definition

All additions are valid CommonMark syntax.

[Rendered](https://github.com/killercup/rfcs/blob/intra-rustdoc-links/text/0000-intra-rustdoc-links.md)

## Revisions

**March 9**: Dropped `<Iterator>`-style links, added implied reference links (https://github.com/rust-lang/rfcs/pull/1946#issuecomment-285419404)

**May 28**: Add more possible extensions, add more syntax for dealing with path ambiguities (https://github.com/rust-lang/rfcs/pull/1946#issuecomment-304504721)

GuillaumeGomez commented:
Awesome! \o/

fbstj commented:
Bikeshed: would it be useful to have an additional `<std::iter::Iterator>` style syntax/marker that desugars to `[Iterator](std::iter::Iterator)` for convenience?

killercup commented:
> an additional `<std::iter::Iterator>` style syntax/marker that desugars to `[Iterator](std::iter::Iterator)`

Hm, I'm wary of adding more cases, and even more so to add magical new syntax. I'd prefer to go with explicit links if you want to set the link text to something other than the full path.

But: If `Iterator` is in scope, `<Iterator>` will be a valid link!

sgrif commented:
This will need a way to disambiguate between functions, traits, etc. It's completely valid to have a struct and function with the same name. Perhaps we could use URIs for this? `function://path::to::fn`, `struct://path::to::struct`, etc.

Also should this specify whether `super` works or not? Are the paths always relative to the crate root? Is there ever a case where it would matter whether the link is based on where the struct appears in the documentation vs where it was defined?

kennytm commented:
Previous discussions 🚲 

* #1661 (*Relative link syntax for rustdoc*, closed)
    * `<::/std/cmp/trait.Ord.html>`
* #792 (*Links to Rust items in documentation text*, active issue) 
    * `<<<std::cmp::Ord>>`
* https://internals.rust-lang.org/t/2720 (*Pre-RFC: standard form to reference things*)
    * `{std::cmp::Ord#cmp}`
    * `[cmp]{@link std::cmp::Ord.cmp}`
    * `<rustdoc:std::cmp::Ord>`
    * `<rustdoc://std::cmp::Ord>`
    * `<rust:std::cmp::Ord>`
    * <code>&grave; std::cmp::Ord&grave;</code> (note leading space)
    * `[std::cmp::Ord]`
* https://internals.rust-lang.org/t/968 (*Rustdoc: Link to other types from doc comments*)
    * <code>&grave;&grave;std::cmp::Ord&grave;&grave;</code>
    * `<rust://std/cmp/Ord.trait>`

frewsxcv commented:
Relevant rust-lang/rust issues: https://github.com/rust-lang/rust/issues/36417, https://github.com/rust-lang/rust/issues/38400

kennytm commented:
@sgrif It only needs to distinguish between values (functions, constants) and types (structs, enums, traits, modules, type-aliases, primitives). Maybe:

* `<syntax::ptr::P>` — First search for type-like items. If not found, search for value-like items
* `<syntax::ptr::P()>` — *Only* search for functions.
* `<std::stringify!>` — *Only* search for macros.

killercup commented:
> This will need a way to disambiguate between functions, traits, etc.

Oh, I had completely missed that this is valid (damn you, Rust!). Is there are an existing way to disambiguate this in paths? Or is it always resolved by the position it appears in?

Playing around on the playpen a bit, it seems that (assuming all have the same name):

- unit and tuple structs conflict with functions
- unit and tuple structs conflict with enums
- regular structs, enums, and traits conflict with each other (but not functions)

I think @kennytm idea with a `()` suffix to disambiguate functions (<https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284719684>) works well. I propose it is only needed when the non-suffixes link is ambiguous. (I assume most code bases follow Rust's style lints and use snake_case functions names and PascalCase for type/trait names.)
Macros should alway use `!` suffix.

> Are the paths always relative to the crate root?

Good point, that's totally missing from the RFC. Path are relative to the position of _item_ in whose documentation they appear. (See complex example for relative paths.)

> Previous discussions

Thanks, @kennytm! Reading <https://github.com/rust-lang/rfcs/pull/1661#issuecomment-230701393>, it only feels right to cc @JohnHeitmann.

frewsxcv commented:
There is talk about [extending CommonMark to support syntactic extensions](https://talk.commonmark.org/t/generic-directives-plugins-syntax/444), but it looks like progress has stalled a bit.

If we're going to design our own hyperlink destinations (as proposed in this PR and the comments above), I'd strongly recommend staying within the bounds of what is acceptable for a ["link destination" as defined by CommonMark][1].

[1]: http://spec.commonmark.org/0.27/#link-destination

kennytm commented:
How will cross-crate links be handled? Like, 

* linking from `num-traits` to `<std::u32>` (linking to standard library)
* linking from `num-complex` to `<num_traits::One>` (crates in the same workspace)
* linking from `num-bigint` to `<serde::Serialize>` (third-party crates, and only linked when a feature is defined)

----

BTW there is also [this thread on talk.commonmark.org](https://talk.commonmark.org/t/what-should-the-rust-community-do-for-linkage/2141) discussing the issue from the Markdown side, but it is pretty inconclusive (one suggestion to `api:std.cmp.Ord` and one to `{std::cmp::Ord}`).

killercup commented:
> How will cross-crate links be handled?

@kennytm, rustdoc is already able to generate links to other crates' types, and I would assume we can use this mechanism. For std (and core), it links to an official URL, e.g. <https://doc.rust-lang.org/nightly/core/fmt/trait.Debug.html>. For other external crates, it links to the docs it generated for the crate itself.

killercup commented:
Updated the RFC with sections on

- Resolving paths
- Path ambiguities
- Linking to external crates

- - - -

I don't think we need to use a potential Markdown extension for this.

> If we're going to design our own hyperlink destinations (as proposed in this RFC and the comments above), I'd strongly recommend staying within the bounds of what is acceptable for a "link destination" as defined by CommonMark.
> – <https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284722655>

I think Rust paths already comply with that. Even the `()` suffix should not be problem, as the fall into part (b) of:

> […] includes parentheses only if (a) they are backslash-escaped or (b) they are part of a balanced pair of unescaped parentheses […]
> – <http://spec.commonmark.org/0.27/#link-destination>

steveklabnik commented:
/cc @jonathandturner who has been working on a rough rustdoc re-write

durka commented:
> Good point, that's totally missing from the RFC. Path are relative to the position of _item_ in whose documentation they appear. (See complex example for relative paths.)

This is a problem for docs that get cross-included by `Deref`.

killercup commented:
> This is a problem for docs that get cross-included by Deref. (<https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284794878>)

Can you give an example? I found [`serde::bytes::ByteBuf`](https://docs.rs/serde/0.9.11/serde/bytes/struct.ByteBuf.html#method.deref) which has a `Deref` impl, and the docs from `trait Deref` itself are rendered.

Including docs from traits should lead to the docs being evaluated relative to the trait item they are defined on, i.e. relative to `core::ops::Deref`, not `serde::bytes::ByteBuf`.

nrc commented:
@killercup, @sgrif, @kennytm 

> This will need a way to disambiguate between functions, traits, etc. It's completely valid to have a struct and function with the same name. Perhaps we could use URIs for this? function://path::to::fn, struct://path::to::struct, etc.

> It only needs to distinguish between values (functions, constants) and types (structs, enums, traits, modules, type-aliases, primitives)

Rust has three namespaces - values, types, and macros. Everything is in one of those three and you don't need any more namespaces than that (it is a bit more complicated because some things can be in more than one namespace, tuple structs, iirc can be used both as type names and as function names). There are already alternate URLs in Rustdoc that use this scheme, unfortunately via redirects (I would love to deprecate the crazy ones with `trait` etc in the name, but there were back compat worries), these are easy to synthesise, e.g., `https://doc.rust-lang.org/nightly/std/path/Path.t.html` for `https://doc.rust-lang.org/nightly/std/path/struct.Path.html`, so you don't need to know the random rustdoc categorisation of a thing, only its namespace.

steveklabnik commented:
Thanks @killercup ! Here's some thoughts....

1. For me, the priority here is to make sure that we're within regular Markdown rules. It appears that we are. 👍 
2. A significant proposal from before that doesn't seem to be addressed here is a `rust://` URI. Putting this in alternatives at least should happen; it's not clear to me that we shouldn't do this.
3. In general, an RFC like this is largely split between two things: how do I define the links, and how does rustdoc actually figure them out. The former is definitely RFC-worthy for me, but I wonder if the latter can't be avoided as some sort of "you use the canonical path for the item linking to", and then free rustdoc to figure that out itself. 
4. In contradiction to point 4, "how does rustdoc figure this out" is a pretty significant pain point, and the hardest part of the RFC, IMHO.
5. @jonathandturner and I (among others at various times) have wondered if rustdoc in the future shouldn't be built on top of RLS. If so, that split would mean "rustodc proper asks RLS what the URL is", I _think_. 
6. thoughts about migrating to this new style might be helpful, I wonder how hard it would be to create a tool to convert existing links like this.

That's all (lol, I feel like it's a lot) I've got at the moment, I'm sure I'll have more eventually...

durka commented:
@killercup 

For example, there is a link from `str::len` to `char`, but this link is also present (and broken) on the `String` page because it is a relative link. There are open issues about this.

It's definitely solvable, but it could require "absolute" (i.e. crate-relative) paths in some places.

nrc commented:
> If so, that split would mean "rustodc proper asks RLS what the URL is", I think.

FWIW, the RLS already creates concrete URLs from paths like this so that when you have a reference to something from std, you can go to the rustdoc page. It might not work exactly how you need it to for this application though.

kennytm commented:
How should Rustdoc know that `<P>` is an HTML tag or a link to a `P` item?

jonathandturner commented:
That's great!

Thanks for the ping.  I think for a "rustdoc 2.0" idea, we're still a little ways off, yet.  Yeah, it will probably consume metadata rather than trying to work against the compiler directly, but there's still a fair bit of research we need to do to explore what we can do there.  That said, I'm currently doing the `rls` stuff, and what we're doing there will help make a "rustdoc 2.0" possible.

I say go for it and keep improving this rustdoc.  It just sets the bar that much higher for the next one (which is a good thing!)

eddyb commented:
@jonathandturner IMO, instead of trying to serialize piecewise, we could have sources for everything, which combined with warm incremental recompilation caches would allow introspecting entire crates.

Even if we don't get incremental reparsing, we could still skip parsing and expansion for, say, a `libstd` that was compiled once and distributed, so you get almost instant access to the HIR and type information.
To keep it slim, we could drop anything that's not exported (i.e. function bodies' MIR) from the caches.

solson commented:
> 2. [Iterator][iter], and somewhere else in the document: [iter]: std::iter::Iterator

@killercup Markdown has another trick that would make this even sweeter at the use sites:

`[Iterator]`, and somewhere else in the document: `[Iterator]: std::iter::Iterator`

llogiq commented:
The 'easy' alternative is to auto-link all items in scope, as in `[String]`. This is trivially supported by pulldown-cmark with extending the links map.

killercup commented:
Okay, let's answer ALL the comments! After the reply to <https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284904948> got a bit longer, I rearranged my other replies a bit. If I forgot something, please ping me.

> How should Rustdoc know that <P> is an HTML tag or a link to a P item? (https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284904948)

Very good question! Sadly one which made me read the CommonMark spec. And the section on [Autolinks](http://spec.commonmark.org/0.27/#autolink) _sadly_ [says](http://spec.commonmark.org/0.27/#uri-autolink):

>  URI autolink consists of `<`, followed by an absolute URI not containing `<`, followed by `>`. It is parsed as a link to the URI, with the URI as the link’s label.
>
> An absolute URI, for these purposes, consists of a scheme followed by a colon (`:`) followed by zero or more characters other than ASCII whitespace and control characters, `<`, and `>`. If the URI includes these characters, they must be percent-encoded (e.g. `%20` for a space).
>
> For purposes of this spec, a scheme is any sequence of 2–32 characters beginning with an ASCII letter and followed by any combination of ASCII letters, digits, or the symbols plus (”+”), period (”.”), or hyphen (”-”).

Which means it is _not_ the same as a [link destination](http://spec.commonmark.org/0.27/#link-destination). Possible to circumvent for the HTML tag ambiguity -- which _only_ stems from the fact that HTML tag names [cannot](http://spec.commonmark.org/0.27/#tag-name) contain colons (and thus, don't work with XML namespaces). But, for some reason I probably missed, [Example 578](http://spec.commonmark.org/0.27/#example-578) of the spec shows that `<m:abc>` will be rendered as neither HTML tag (colon), nor autolink (no idea why).

So, what does it mean for this RFC? The `<nice::short::fancy::Autolinks>` I proposed are not valid CommonMark.

Are they valid [Markdown](https://daringfireball.net/projects/markdown/syntax#autolink)? That depends on your definition of URL. The consensus is that URLs need to start with a schema, though, so `<String>` is not valid.

Which brings us to:

> A significant proposal from before that doesn't seem to be addressed here is a rust:// URI. (https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284827670)

This solves the autolink problem, `rust://String` is valid Markdown. To me, though, adding a URL scheme makes everything look a bit more complicated than necessary. If we can omit it, doc comments will look much cleaner.

Assuming we are okay with interpreting the CommonMark spec a bit more lax, let's have a look at what the [official renderer](http://spec.commonmark.org/dingus/) actually produces:

```md
Look at me, using a <String>!
```

becomes

```html
<p>Look at me, using a <String>!</p>
```

and

```md
Look at me, using a <std::string::String>!
```

becomes 

```html
<p>Look at me, using a <a href="std::string::String">std::string::String</a>!</p>
```

We can see that the first example produces invalid HTML, so nobody should write this anyway. The second one sees `std::string::String` as a valid URL (`std:` is a good a URL scheme!).

This seems like a good point to reply to:

> In general, an RFC like this is largely split between two things: how do I define the links, and how does rustdoc actually figure them out. The former is definitely RFC-worthy for me, but I wonder if the latter can't be avoided as some sort of "you use the canonical path for the item linking to", and then free rustdoc to figure that out itself. (https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284827670)

I'm not sure I understand what you mean by the latter part, @steveklabnik. Do you want people to always have to write the "canonical path"? Is "canonical path" the "absolute path", i.e., `lorem::ipsum::Dolor` instead of `Dolor` (when inside the `lorem::ipsum` module)? That could potentially make things simpler, as rustdoc then does not need to resolve relative paths. (But, it would still need to _render_ URLs relative to the current page.)

Requiring that paths have _at least_ two segments can also ensure that all paths are actually valid URLs -- the first segment plus one column are a valid schema (see above). I'm not sure this is an ergonomic solution, though. People will have less of a problem remembering to write `std::collections::HashMap` to refer to a type that is not in scope, but will wonder why `<HashMap>` does not work when they have a `use std::collections::HashMap;` just above the commented item. (`<self::HashMap>` would work.)

I'm unsure how to proceed further. I would like to keep using standard Markdown, but I also _really_ like the short autolink form.

- - -

> In contradiction to point 4, "how does rustdoc figure this out" is a pretty significant pain point, and the hardest part of the RFC, IMHO. (https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284827670)

Do you mean this should be specified in the RFC? This is pretty much like an implementation detail of rustdoc. I mean, the basic algorithm is,

1. Resolve the path using Rusts module system
	- maybe canonicalize first if that's easy and helpful
	- error out if the path is invalid/cannot be resolved
2. Convert this Rust path to the URL of the page (possibly with fragment specified, e.g. `std/iter/trait.Iterator.html#method.fuse`)
3. Turn this URL into one relative to the page this currently being rendered (e.g. `../iter/trait.Iterator.html#method.fuse`)

By doing the last step(s) in the context of a concrete rustdoc page being rendered, it will generate correct, relative links. This should also address <https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284842869>.

> Stuff about using RLS (<https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284827670>, <https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284827670>)

Could be useful, but also an implementation detail. I don't want this to be a design we chose because the implementation was easy, but one that is in itself useful (also addresses <https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284960338>).

If there are specific limits of what we _can_ do, I'm all ears, of course!

- - -

> Markdown has another trick […] (https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284953009)

Yes, I'm using this quite heavily in the Complex Example section :)

killercup commented:
Seeing autolinks become difficult to combine with standard Markdown, one way to get a similar syntax is to reevaluate the proposal of automatically generating link reference definitions, which would allow writing `Look at me, using a [String]` (or ```… [`String`]```, or `… [std::string::String]`).

It might also be a good idea to have a look at the existing code bases to see what actually gets linked to. This be skewed to some degree of course, as cross linking is currently not that easy.

> thoughts about migrating to this new style might be helpful, I wonder how hard it would be to create a tool to convert existing links like this. (https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284827670)

Assuming Rust path ↔ Rustdoc URL is bijective, it's only a matter of getting the right Markdown tokens to convert between the two. (Re)writing nicely formatted doc comments might be a bit harder, though.

durka commented:
@llogiq there should still be a way to link to things that aren't in scope, or are ambiguous (e.g. `std::io::Result` and `std::result::Result`).

durka commented:
@killercup is `<rs:String>` any more palatable? Or multiple schemes, like `<type:String>`/`<macro:println>`.

kennytm commented:
@killercup 

> For purposes of this spec, a scheme is any sequence of 2–32 characters beginning with an ASCII letter and followed by any combination of ASCII letters, digits, or the symbols plus (”+”), period (”.”), or hyphen (”-”).

Sadly this means `<num_traits::One>` won't be valid either because of the `_`.

> **A compromise might be use use rust: as a scheme (without the double slashes).** `<rust:String>` is a valid autolink. Using this scheme for autolinks only seems unelegant, though.

Perhaps allow both in auto-link and normal links, i.e. `[Iterator](rust:Iterator)` and `[Iterator](Iterator)` should be the same. 

I don't like the lack of `//` though, because `rust:Iterator` and `rust::Iterator` look too similar.

kennytm commented:
I think @llogiq's comment https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284960338 is a viable alternative which does not break CommonMark. Conceptually, rustdoc could generate links of all visible items and append them to the end of the generated markdown like

```md
[Clone]: https://doc.rust-lang.org/nightly/core/clone/trait.Clone.html
[`Clone`]: https://doc.rust-lang.org/nightly/core/clone/trait.Clone.html
[One]: ../identities/trait.One.html
[std::clone::Clone]: https://doc.rust-lang.org/nightly/core/clone/trait.Clone.html
[num_traits::One]: ../identities/trait.One.html
...
```

So one just write normal markdown like ``The first argument of the [num_traits::pow()] function needs to implement the [`Clone`], [`One`] and [`Mul<Self>`] traits...``. 

1. `[Iterator](std::iter::Iterator)` becomes `[Iterator][std::iter::Iterator]`.
2. `[Iterator][iter]` becomes impossible 😢.
3. `<std::iter::Iterator>` becomes `[std::iter::Iterator]`.

For sure, the list needs to be dynamically generated, otherwise it will blow up the markdown parser.  The parser will need to allow a hook to detect the input which contains `[invalid_reference]` or `[link][invalid_reference]` or `![image][invalid_reference]`. The callback function will determine if `invalid_reference` can refer to a Rust item. If yes, the callback should replace the invalid reference by the URL to the corresponding item.

killercup commented:
@kennytm's last proposal (https://github.com/rust-lang/rfcs/pull/1946#issuecomment-285078390) sounds good, but I'd tweak it like this:

1. ✅ `[Iterator](std::iter::Iterator)`
2. ✅`[Iterator][iter]`, and somewhere else in the document: `[iter]: std::iter::Iterator`
3. ✅ `[Iterator]`, and somewhere else in the document: `[Iterator]: std::iter::Iterator` (this takes precedence before the following rules)
4. ✅ `[Iterator]` and `[std::collections::HashMap]` work, when `Iterator`/`std::collections::HashMap` are resolvable paths, the `[Iterator]: std::iter::Iterator` is implied (lazily added by rustdoc)
4. ✅ ```[`Iterator`]``` works just like 4
5. ❌ `<std::iter::Iterator>` does **not** work

killercup commented:
I'd like to not introduce a new URL scheme for this, but:

> is <rs:String> any more palatable? (https://github.com/rust-lang/rfcs/pull/1946#issuecomment-285076632)

Yeah, I think `rs:` is nicer than `rust:` or `rustdoc:`.

> Or multiple schemes, like <type:String>/<macro:println>.

Hm, if we go down that route (I don't want to), it should not be `type:`, but `struct:`, `enum:`, `trait:`, etc.

ssokolow commented:
> Hm, if we go down that route (I don't want to), it should not be type:, but struct:, enum:, trait:, etc.

Given that there's enough of a problem with programmers being rushed, burned out, or just plain lazy as-is, I'd follow Sphinx's lead in that any system which has multiple "roles" used for cross-referencing (eg. `:ref:`, `:py:mod:`, etc.) should also have an `:any:` role which'll just do what you mean™ as long as you exercise good naming discipline.

Sphinx did a pretty good job of mitigating laziness there by combining that aspect of reStructuredText's extensible-by-design syntax with the "default role" feature to allow you to map `` `identifier` `` to `` :any:`identifier` `` (with ReST using ``` ``this`` ``` for inline code literals instead) and then `:any:` to whatever unique identifier you intended.

(That said, Sphinx *really* dropped the ball by embracing "we're a replacement for LaTeX, not epydoc". To this day, the "inject bits of autodoc into your existing document"-oriented approach to their autodoc frequently sends me diving into the source to figure out something a Python developer thought was too obvious to autodoc. I'll take "the search box is your cross-referencing tool" over "the source code is your API docs" any day.)

killercup commented:
Alright, I updated the RFC text.

- **Autolinks are out**. Too much pain with too many edge cases.
- **Implied link reference definitions are in**! This means `[HashMap]` works (but a manual link reference definition takes precedence.)

I hope this makes @GuillaumeGomez, @llogiq, and every one else very happy :)

steveklabnik commented on text/0000-intra-rustdoc-links.md:
```diff
@@ -0,0 +1,389 @@
+[CommonMark forum]: https://talk.commonmark.org/t/what-should-the-rust-community-do-for-linkage/2141
+[Guillaume Gomez]: https://github.com/GuillaumeGomez
+
+## Standard-conform Markdown
```
conforming

fbstj commented:
I don't know if this is at all relevant, but github have just released (or at least publicised) the [GitHub Flavored Markdown spec](https://github.github.com/gfm/) and their fork of [cmark](https://github.com/github/cmark)

steveklabnik commented:
I've re-read this, great work @killercup ! I'm currently 👍 

Regarding the Unresolved Quesitons:

> Is it possible for Rustdoc to resolve paths? Is it easy to implement this?

This is a _huge_ question.

> There is talk about switching Rustdoc to a different markdown renderer (pulldown-cmark). Does it support this? Does the current renderer?

This is in-progress https://github.com/rust-lang/rust/issues/38400

Given that it's compliant with CommonMark, or at least, should be, if this is compliant with CommonMark, we should be good, I'd imagine.

llogiq commented:
I had a look at pulldown_cmark's source. It currently uses a `links: HashMap<..>` – but I believe it should be easy enough to allow for a lookup extension to compute the actual links on demand. Then it'd be a matter of gathering all items in scope and synthesizing the respective links.

frewsxcv commented on text/0000-intra-rustdoc-links.md:
```diff
@@ -0,0 +1,389 @@
+
+Ideally, Rustdoc would be able to recognize Rust path syntax,
+and if the path cannot be resolved,
+print a warning (or an error).
```
In the case of a warning, do docs still get generated? If so, will there be a broken link?

GuillaumeGomez replied:
Docs generate urls, wether they're valid or not. rustdoc won't generate docs if the code doesn't compile (so warnings are fine).

Mark-Simulacrum commented:
Regarding the unresolved questions:

> Is it possible for Rustdoc to resolve paths? Is it easy to implement this?

Well, at the very least https://github.com/rust-lang/rust/pull/34655 implemented a function on resolver that should allow resolving paths with an arbitrary module's lexical scope, which is close to what we need, though ideally it'd be any scope, not just module-level scope.

nrc commented:
Another round of comments:

* I'm still strongly in favour of the idea - I think Rustdoc needs this
* I like the format for specifying links, seems like a good balance between flexibility and sticking to the spec, etc.
* I don't think it is worth going over the implementation questions in the RFC - its seems to me that it is not hard to make it work and that we just need to plugin to name resolution. Exactly how we do that depends on how we implement Rustdoc (esp if we rewrite using RLS tech).
* I'm a bit wary about special casing inter-crate links to the standard libraries without thinking of a general mechanism. In particular - how do we choose a URL to link to? I don't think there are hard questions here - we should presumably be using the version of the docs that correspond to the version of the compiler being used, but how do we get that info to Rustdoc? (Especially in a future where Rustdoc does not link the compiler). How can the user override the default URLs? Or link to docs.rs rather than relative links to dep docs, etc.
* I strongly feel that the 'path ambiguities' section needs fixing - it is clear how the language works here, and we should follow that, not some approximation. There are three namespaces - types, values, and macros, authors should be able to specify one of those in case of ambiguity.
* Will we allow paths to fields, variants, and associated items? It seems these should be fairly straightforward. I think using 'qualified self' paths to methods and other impl items should be allowed too - although I imagine this might be harder to implement, I don't see any trickiness in adding them to the RFC because they have clearly defined semantics.

killercup commented:
Thanks, @nrc! These are great points. I have updated the RFC and it a few new things.

- I have expanded the [path ambiguity section](https://github.com/killercup/rust-rfcs/blob/82ea2bc6ef0b9f00c383558cf11eecc46556ea8e/text/0000-intra-rustdoc-links.md#path-ambiguities) to propose explicit syntax to address the item. Let me know what you think of this, I just came up with the syntax and would be happy to bikeshed :)
  - *tl;dr* `see trait [FOO]`, `see struct [FOO]`, `see [const FOO]`, `see [static FOO]`, `see [FOO()]`, `see [FOO!] macro`.
  - I wanted to include a "If there is only one possible item, you can just write [FOO]", but I'm not sure about backward compatibility—what if a crate adds a `pub fn FOO` and you were linking to a `trait FOO`?
- I have added the following under "Possible Extensions" (if you all think this is good, we can move it into the Detailed Design section)
  - [Linking to fields and variants](https://github.com/killercup/rust-rfcs/blob/intra-rustdoc-links/text/0000-intra-rustdoc-links.md#linking-to-fields-and-variants)
  - [Linking to external documentation](https://github.com/killercup/rust-rfcs/blob/intra-rustdoc-links/text/0000-intra-rustdoc-links.md#linking-to-external-documentation) with a proposal how to generalize links to externally rendered docs
  - Slightly changed [Linking to associated items](https://github.com/killercup/rust-rfcs/blob/intra-rustdoc-links/text/0000-intra-rustdoc-links.md#linking-to-associated-items)

QuietMisdreavus commented on text/0000-intra-rustdoc-links.md:
```diff
@@ -0,0 +1,465 @@
+it may be necessary to use fully-qualified paths,
+like `See the [<Foo as Bar>::bar()] method`.
+We have yet to analyze in which cases this is necessary,
+and this syntax is currently not described in [the reference's section on paths][ref-paths].
```
While it's not in the reference under paths, thanks to [rust-lang-nursery/reference#45](https://github.com/rust-lang-nursery/reference/pull/45), there's now [a mention of fully-qualified paths under expressions](https://github.com/rust-lang-nursery/reference/blob/d9243090c118439c69d325fc0509938809142666/src/expressions.md#disambiguating-function-calls)!

nrc commented:
re the path ambiguities, iiuc, your proposal is that the default is to point to the type if there are names in multiple namespaces? My preference would be that no qualifier only works if there is a single namespace and is an error (or at least a warning otherwise). Then you can qualify with `type`, `value`, or `macro` to be explicit about the namespace. Using existing keywords (`static`, `const`, etc.) seems annoying for tools because you need to know the kind of the definition, not just the namespace. I guess this would be acceptable if you could always use any of those qualifiers for any kind of def (e.g., I could use `const` even if the def is a function), but that still leaves the problem of disambiguating to the type namespace.

killercup commented:
> My preference would be that no qualifier only works if there is a single namespace

I'm not sure I follow. Can you give an example of when there can only a single namespace? My concern is that docs linking to external crates should not break when the crates add (public) items.

> Using existing keywords (static, const, etc.) seems annoying for tools

That's a good point. I was mostly thinking about making the default link texts read nicely in English, without considering any implementation complexity. (And we can of course bikeshed the actual syntax.) Do you think this is a hard problem or something that is in the end just a mapping of prefixes to namespaces?

nrc commented:
We discussed this today in the dev-tools meeting (https://github.com/nrc/dev-tools-team/blob/master/minutes/meeting%20notes%202017-06-05.md). We concentrated on the namespaces detail. Other than that, we think this is ready for RFC.

Outstanding questions:
* syntax for disambiguating the namespace
* should a non-disambiguated namespace have a default or be an error? Personally, I think that is must be an error

For the syntax, my constraints are:
* a tool or user writing a link should not have to know the kind of the item being referred, only the namespace
* the reference must only depend on the namespace
* the name can be disambiguated into any namespace

To adapt @killercup's proposal: the `const`, `static`, or `fn` prefixes indicate the value namespace. The `macro` prefix indicates macro namespace. Any of `type`, `trait`, `mod`, `struct`, etc. indicate the type namespace. Importantly one could define a module `foo` and refer to it using `struct ::foo`.

Alternative: `type`, `value`, `macro` as possible prefixes. Or `t`, `v`, `m`.

There are probably better ideas too.

killercup commented:
@nrc, sorry for the delay. I just updated the RFC to expand a bit on the namespace disambiguation section.

My proposal is still to go with a default namespace (types), and to suffix macros with `!`, and function with `()` as disambiguation. My goal with this is to make writing the docs, and reading them in code feel as natural as possible. Namespace prefixes (type, value, macro or even t, v, m) seem cryptic and force the author to know/think about something that should not directly matter when writing documentation (especially compared to "just copying" the Rust keyword used to create the item in code). Additionally, my intuition is that types are linked to most often (with functions in second place), so making them as short as possible seems good.

→ It would be awesome if someone could gather some empirical evidence for/against this!
→ I'm fine with bikeshedding this a bit more :)

(Other than that, the RFC now also allows prefixes for items in type namespace, and mentions that it's okay if tools allow `[enum Foo]` to link to a struct called `Foo`.)

est31 commented:
Little question. What should be the context from which `[item]` name resolval starts from? As in, should it be the inside of the item we document, or the outside?

Example:

```rust
use baz::bar;

/// `[bar]` <--- what bar does this resolve to?
mod foo {
    mod bar;
}

use foo::Iterator

/// `[Iterator]` <--- what "Iterator" does this resolve to?
fn foo() {
    use bar::Iterator;
}
```

killercup commented:
Good question. In addition:

```rust
use foo::Iterator

fn foo() {
    //! `[Iterator]` <--- what "Iterator" does this resolve to?
    use bar::Iterator;
}
```

QuietMisdreavus commented:
From discussion on IRC: If we determine what items are in scope (or more closely in scope) via whether the doc comment is applied from inside or outside the item in question, it may be necessary to extend the attributes system in the compiler/rustdoc/analysis to state whether an attribute is an external or internal form. As-is, I don't think any of those tools can make the distinction.

Otherwise, I like the new changes! `(^^)b`

nrc commented:
> Little question. What should be the context from which [item] name resolval starts from?

I think we should treat resolution based on whether it is inside or outside, i.e., `///` starts from the parent`, `//!` starts from inside the documented item. This will need some more info stored in the AST for attributes, but it is not a major change.

nrc commented:
@killercup "we do not require the tool resolving the links ..." - I think this should be stronger and state that any item kind is acceptable

killercup commented:
Updated the RFC to include <https://github.com/rust-lang/rfcs/pull/1946#issuecomment-307826090>, <https://github.com/rust-lang/rfcs/pull/1946#issuecomment-307827544>, <https://github.com/rust-lang/rfcs/pull/1946#issuecomment-308646046>, as well as <https://github.com/rust-lang/rfcs/pull/1946#issuecomment-308646405>.

QuietMisdreavus commented on text/0000-intra-rustdoc-links.md:
```diff
@@ -0,0 +1,521 @@
+relative to the item in whose documentation they appear.
+Specifically, when using inner doc comments (`//!`, `/*!`),
+the paths are resolved from the inside of the item,
+while regular doc comments (`///`, `/**`) start from the "parent scope.
```
Not sure what that quote is supposed to be.

killercup replied:
a typo :) I'll go over it again and fix some

nrc commented:
@rfcbot fcp merge

killercup commented:
From my end, this looks is to be merged. If anyone wants to talk about wonderful new syntax alternatives… it's your last chance! :)

@rfcbot reviewed

jonathandturner commented:
@rfcbot reviewed

ollie27 commented:
How would you create links to primitive types like arrays, pointers, slices, and tuples?

killercup commented:
@ollie27, that's a good question! As they are language features, have special names as well as rustdoc pages, and are not that many, I think this RFC should not include a special way to link to primitives. So, the we'll need to link to the std docs, e.g. with `[array](https://doc.rust-lang.org/1.18.0/std/primitive.array.html)`.

ssokolow commented:
@killercup 

I strongly disagree. Given how often I see people not bothering to mark up their cross-references in various languages, it's quite likely that kind of verbosity for cross-referencing primitives will encourage people to just not bother (especially when they're used to the more compact syntax for non-primitives).

When you're new enough to the language to not yet be familiar with the APIs the primitives expose is when you need easy and ubiquitous cross-referencing the *most*.

...not to mention that the use of absolute URLs to cross-reference primitives in rustdoc feels like it runs counter to the spirit of having a stability promise for the language being documented.

ollie27 commented:
A lot of the broken links on https://doc.rust-lang.org/nightly/std/string/struct.String.html are links to primitive types so it would be a shame not to be able to fix them. How about adding another "namespace" which uses the names rustdoc uses for the primitives so things like `[primitive u8]` and `[primitive slice.sort]` would work?

nrc commented:
@rfcbot cancel

nrc commented:
@rfcbot fcp merge

killercup commented:
I see your point. That is a valid use case. I'm okay with adding it. I have
a few questions though:

Is 'primitive' a namespace rust knows, or actually just a thing that
rustdoc uses for various built-in stuff? (I assume the latter.) Can we easy
generate these links?

We are linking to types by default, and for many primitives there are also
types with the same name (stuff like i32). This leads me to assume that
many people will use the short, easy way and link to the types. Is that
okay? Do these types all link to their respective primitives page?

Nick Cameron <notifications@github.com> schrieb am Di. 20. Juni 2017 um
00:26:

> @rfcbot <https://github.com/rfcbot> fcp merge
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/1946#issuecomment-309590961>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABOX5agiIAzYbggrNUjdiaCrStH6tNxks5sFvWWgaJpZM4MVWGT>
> .
>

est31 commented:
> for many primitives there are also types with the same name (stuff like i32)

i32 there are two doc pages, one for a [module](https://doc.rust-lang.org/stable/std/i32/index.html) `std::i32`, and one [for the primitive type](https://doc.rust-lang.org/stable/std/primitive.i32.html). There has been talk in the lang team to get rid of the module and replace it with associated consts (cc @eddyb ), so in future there might only be the primitive type.

ollie27 commented:
I'm pretty sure rustc special cases the primitive types, I know rustdoc does. rustdoc already has code specifically for generating links to the primitive types so I doubt there will be any issues.

As long as there are no types or module named `i32` in scope is seems reasonable for `[i32]` to link to the primitive page.

nrc commented:
@rfcbot fcp cancel

nrc commented:
@rfcbot fcp merge

japaric commented:
@rfcbot reviewed

michaelwoerister commented:
@rfcbot reviewed

QuietMisdreavus commented:
@rfcbot reviewed

nrc commented:
ping @brson for fcp approval

killercup commented:
Just came up:

    `[Iterator]`

should probably not be rendered as a link inside a code block.

Manishearth commented:
We're proposing some amendments in https://github.com/rust-lang/rfcs/pull/2285

Manishearth commented:
@killercup @QuietMisdreavus should we turn the error into a warning ? I don't like the idea of crate docs breaking because someone introduced a name into a glob import
